[Bellatrix] introduces optimisticSync-aware SyncStateTracker

tbenr opened the pull request:
## PR Description
makes `SyncStateTracker` optimisticSync-aware.

## Fixed Issue(s)
fixes #4783


## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

tbenr committed:
introduces optimisticSync-aware SyncStateTracker

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java:
```diff
@@ -64,6 +66,10 @@
 
+  private final Subscribers<OptimisticSyncSubscriber> optimisticSyncSubscribers =
+      Subscribers.create(true);
+  private final AtomicBoolean optimisticSyncing = new AtomicBoolean(false);
```
This is only ever accessed from the fork choice thread so doesn't actually need to be thread safe. That said, the way the compareAndSet winds up working to only deliver events when the value changes is quite neat code so may be worth sticking with the AtomicBoolean for that simplicity...

tbenr replied:
i was bouncing between boolean and AtomicBoolean, it isn't that evident that `onBlock` is called only from the fork choice thread... And TBH i'm still not convinced :)

ajsutton replied:
We're pretty screwed if it's called from any other thread. :)

The key is that it's called via:
```
.thenApplyAsync(
            payloadResult ->
                importBlockAndState(...),
            forkChoiceExecutor);
```

so it's the `forkChoiceExecutor` that will execute it and that ensures it's always on the event thread.

tbenr replied:
ahh yes I missed that!

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java:
```diff
@@ -311,6 +317,14 @@ private BlockImportResult importBlockAndState(
+              .orElse(false)) {
+        optimisticSyncSubscribers.deliver(
+            OptimisticSyncSubscriber::onOptimisticSyncingChanged, true);
+      }
```
I wonder if it would be simpler to detect when the chain head is optimistic by adding that info to the `ForkChoiceState` returned by `ForkChoiceStrategy.getForkChoiceState` which is called from `ForkChoice.notifyForkChoiceUpdated`.  We know that `notifyForkChoiceUpdated` is always called when the optimistic chain head changes and it's simple for the protoarray to tell if the head block it chose is optimistically sync'd or not.  Then we'd have a single place in the code that updates `optimisticSyncing` and notifies subscribers when it changes.

Comparing the best block root to the new block parent root makes me a little nervous as `recentChainData.updateChainHead` is async so it's possible we'd miss a switch to optimistic sync with this condition and it only ever detects the first optimistically sync'd block so wouldn't recover.

tbenr replied:
yeah I wasn't proud of the solution, I just put it there to start this discussion :)

now I implemented an inner class holding `ForkChoiceState` and the `isHeadOptimistic` boolean.
I bounced back and forth from this solution and the one adding the boolean directly in the `ForkChoiceState` class.

I ended up with this to avoid adding complexity in the ForkChoiceNotifier area, where this information is not needed, even if it was the right place TBH.

ajsutton replied:
oh sorry read this comment then forgot about it in the jcenter fun and commented about putting the boolean in `ForkChoiceState`.  I think I'd still put it in `ForkChoiceState` to be honest but keep the subscribers where they are in `ForkChoice` so `ForkChoiceNotifier` will stay unchanged.  It means an extra boolean is flowing through with that `ForkChoiceState` instance but I'm ok with that - it makes sense from a design perspective that the fork choice state includes whether the head is optimistic or not and it's only one extra boolean so probably less cost than creating the extra wrapper object.

tbenr replied:
another option is to put it in as Optional<Boolean>, so it becomes empty as soon as it is regenerated by  ForkChoiceNotifier. But let me do the right thing first :)

tbenr replied:
wait, the booster affected my head, i confused `ForkChoiceState` with `ForkChoiceStateData`. putting it here is a nobrainer.

ajsutton commented on sync/src/main/java/tech/pegasys/teku/sync/events/OptimisticSyncStateTracker.java:
```diff
@@ -0,0 +1,20 @@
+
+import tech.pegasys.teku.statetransition.forkchoice.ForkChoice;
+
+public interface OptimisticSyncStateTracker extends ForkChoice.OptimisticSyncSubscriber {
```
nit: This is unused and can be deleted.

ajsutton reviewed (commented):
Looks like it's heading in the right direction generally.

tbenr committed:
derives optimistic head from protoArray
puts ForkChoiceUpdated and OptimisticSyncingChanged in the same method
removes ForkChoice static create method
implements UT

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java:
```diff
@@ -392,13 +387,21 @@ private void reportInvalidBlock(final SignedBeaconBlock block, final BlockImport
-    forkChoiceNotifier.onForkChoiceUpdated(forkChoiceState);
+    forkChoiceNotifier.onForkChoiceUpdated(forkChoiceStateAndIsHeadOptimistic.forkChoiceState);
+    if (optimisticSyncing
+        .map(os -> !os.equals(forkChoiceStateAndIsHeadOptimistic.isHeadOptimistic))
```
nit: Probably worth using a more meaningful name than `os` - `oldValue`?

ajsutton commented on protoarray/src/main/java/tech/pegasys/teku/protoarray/ForkChoiceStrategy.java:
```diff
@@ -451,4 +453,15 @@ public void onExecutionPayloadResult(
+
+  public static class ForkChoiceStateAndIsHeadOptimistic {
+    public final ForkChoiceState forkChoiceState;
+    public final boolean isHeadOptimistic;
```
I think I'd just add `isHeadOptimistic` to `ForkChoiceState` rather than adding a wrapper class.  We have deliberately separated the internal `ForkChoiceState` class we use from the `ForkChoiceStateV1` that defines how it serialises so can add extra information that's useful to us without it affecting the execution engine API.

ajsutton reviewed (approved):
LGTM.

tbenr committed:
move isHeadOptimistic in ForkChoiceState

tbenr committed:
spotlessly

tbenr committed:
cleanup tests
